### ISSUE TITLE
Error message during the fitting (data from FHI-aims)

### ISSUE THREAD

**Issue by DGKang234** [libAtoms/QUIP#263]
Hello professors,

As a practice, I have attempted to fit GAP of (ZnO)2 nanocluster using FHI-aims/ase.

[1] I have generated 4 data set from ase QuasiNewton method.
(total 4 frames = 2:training set, 2:validation set and I append e0 of Zn and O to train.xyz)

Then, 
I did fitting with GAP_2b descriptors:
`! gap_fit energy_parameter_name=energy force_parameter_name=forces do_copy_at_file=F sparse_separate_file=T gp_file=GAP.xml at_file=train.xyz default_sigma={0.008 0.04 0 0} gap={distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform add_species=T n_sparse=10}`

However, it aborts the GAP fitting process with the error message below.
I do not understand what it means.
How should I solve the problem?

```
============== Gaussian Approximation Potentials - Database fitting ============


Initial parsing of command line arguments finished.
Found 1 GAPs.
Descriptors have been parsed
XYZ file read
Old GAP: {distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform add_species=T n_sparse=10}
New GAP: {distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform               n_sparse=10 Z1=30 Z2=30}
New GAP: {distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform               n_sparse=10 Z1=30 Z2=8}
New GAP: {distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform               n_sparse=10 Z1=8 Z2=8}
Multispecies support added where requested

===================== Report on number of descriptors found ====================

---------------------------------------------------------------------
Descriptor: distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform               n_sparse=10 Z1=30 Z2=30
Number of descriptors:                        4
Number of partial derivatives of descriptors: 24
---------------------------------------------------------------------
Descriptor: distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform               n_sparse=10 Z1=30 Z2=8
Number of descriptors:                        16
Number of partial derivatives of descriptors: 96
---------------------------------------------------------------------
Descriptor: distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform               n_sparse=10 Z1=8 Z2=8
Number of descriptors:                        4
Number of partial derivatives of descriptors: 24

========================================  ======================================

Atom species 8 present in teaching XYZ, but not found corresponding isolated representative
Atom species 30 present in teaching XYZ, but not found corresponding isolated representative
SYSTEM ABORT: Determination of e0 was requested to be based on isolated atom energies, but not all  atom types
present in the XYZ had an isolated representative.   
```
Kind regards,

I appreciate your time and help.

DG Kang
PhD student.

**Comment by gabor1** [comment]
The error message is: 

> SYSTEM ABORT: Determination of e0 was requested to be based on isolated atom energies, but not all  atom types
> present in the XYZ had an isolated representative.   

So this means that you did not provide frames in the xyz with isolated Zn and (separately) isolated O. 

when you say "append e0 of Zn and O to train.xyz", what does that mean ? 


-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by DGKang234** [comment]
the "append" term was from this script: (actually it rewrite a file)
```
ase.io.write('train.xyz', out_traj[0::2] + [isolated_Zn] + [isolated_O])
ase.io.write('validate.xyz', out_traj[1::2])
```

- This is the contents of "train.xyz"

```
4
Properties=species:S:1:pos:R:3:forces:R:3 energy=-102238.230154691 pbc="F F F"
Zn       0.00041014       0.00000000      -1.14373494      -0.00042784       0.00000000       1.08046155
Zn      -0.00051811       0.00000001       1.29967352       0.00042867      -0.00000000      -1.08218191
O        1.41767369       0.00000001       0.07841698       0.61022699      -0.00000000       0.00110706
O       -1.41778158       0.00000001       0.07733976      -0.61022781      -0.00000000       0.00061330
4
Properties=species:S:1:pos:R:3:forces:R:3 energy=-102238.355090833 pbc="F F F"
Zn       0.00037341       0.00000000      -1.05154464       0.00004464      -0.00000000      -0.22371149
Zn      -0.00048132       0.00000001       1.20734056      -0.00004492       0.00000000       0.22456384
O        1.49660282       0.00000001       0.07851977      -0.16617998       0.00000000      -0.00045926
O       -1.49671077       0.00000001       0.07737963       0.16618026       0.00000000      -0.00039308
1
Properties=species:S:1:pos:R:3:forces:R:3 energy=-49078.1951382926 free_energy=-49078.1951382926 pbc="F F F"
Zn       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000
1
Properties=species:S:1:pos:R:3:forces:R:3 energy=-2034.82069902266 free_energy=-2034.81298590667 pbc="F F F"
O        0.00000000       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000
```

> |  SYSTEM ABORT: Determination of e0 was requested to be based on isolated atom energies, but not all  atom types
> |   present in the XYZ had an isolated representative.

I have calculated each atom separately and I can see that "train.xyz" contains isolated_Zn and isolated_O.
I still don't see which gives the error.

**Comment by gabor1** [comment]
paste me the relevant parts of the train.xyz

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by DGKang234** [comment]
I appreciate your help.

Here is the conents of the "train.xyz"
```
4
Properties=species:S:1:pos:R:3:forces:R:3 energy=-102238.230154691 pbc="F F F"
Zn       0.00041014       0.00000000      -1.14373494      -0.00042784       0.00000000       1.08046155
Zn      -0.00051811       0.00000001       1.29967352       0.00042867      -0.00000000      -1.08218191
O        1.41767369       0.00000001       0.07841698       0.61022699      -0.00000000       0.00110706
O       -1.41778158       0.00000001       0.07733976      -0.61022781      -0.00000000       0.00061330
4
Properties=species:S:1:pos:R:3:forces:R:3 energy=-102238.355090833 pbc="F F F"
Zn       0.00037341       0.00000000      -1.05154464       0.00004464      -0.00000000      -0.22371149
Zn      -0.00048132       0.00000001       1.20734056      -0.00004492       0.00000000       0.22456384
O        1.49660282       0.00000001       0.07851977      -0.16617998       0.00000000      -0.00045926
O       -1.49671077       0.00000001       0.07737963       0.16618026       0.00000000      -0.00039308
1
Properties=species:S:1:pos:R:3:forces:R:3 energy=-49078.1951382926 free_energy=-49078.1951382926 pbc="F F F"
Zn       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000
1
Properties=species:S:1:pos:R:3:forces:R:3 energy=-2034.82069902266 free_energy=-2034.81298590667 pbc="F F F"
O        0.00000000       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000
```

**Comment by gabor1** [comment]
even though these are not periodic, you need a Lattice property, with a big lattice (bigger than twice the largest GAP cutoff), otherwise gap_fit will not parse these properly. 

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by DGKang234** [comment]
So This means that I have to run the (ZnO)2 in pbc = T T T which has a very big cell.
Do I have to apply the same term on a single atom as well?

**Comment by gabor1** [comment]
you don’t have to rerun it, if you are confident that your DFT (whatever it is, plane wave, nonperiodic etc) gave the right energy, then it’s fine, just tell GAP that it’s in a big cell, so images don’t see each other. 

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by gabor1** [comment]
and yes, I think it’s best to give a big cell to all your configs. the point is that the XYZ reader doesn’t know what "big" means, and when it reads your frame without a cell info, it just makes one up, but that might not be big enough, depending on your GAP cutoff. 

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by DGKang234** [comment]
It worked!
I appreciate your help.

DG Kang

**Comment by DGKang234** [comment]
For the following error message during the GAP fitting:
>  SYSTEM ABORT: Determination of e0 was requested to be based on isolated atom energies, but not all  atom types
>  present in the XYZ had an isolated representative.


solution:
>  even though these are not periodic, you need a Lattice property, with a big lattice (bigger than twice the largest GAP 
>  cutoff), otherwise gap_fit will not parse these properly.

Add lattice vector component (Lattice="x x x x x x x x x") at the front of the Properties in 2nd line of  "train.xyz"
**N.B.** cell size should be much bigger than the actual system if you are fitting a cluster/molecule to make sure they only communicate with each other infinitesimally.

e.g.
```
4
Lattice="10.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 10.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-102238.230154691 pbc="T T T"
Zn       0.00041014       0.00000000      -1.14373494      -0.00042784       0.00000000       1.08046155
Zn      -0.00051811       0.00000001       1.29967352       0.00042867      -0.00000000      -1.08218191
O        1.41767369       0.00000001       0.07841698       0.61022699      -0.00000000       0.00110706
O       -1.41778158       0.00000001       0.07733976      -0.61022781      -0.00000000       0.00061330
4
Lattice="10.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 10.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-102238.355090833 pbc="T T T"
Zn       0.00037341       0.00000000      -1.05154464       0.00004464      -0.00000000      -0.22371149
Zn      -0.00048132       0.00000001       1.20734056      -0.00004492       0.00000000       0.22456384
O        1.49660282       0.00000001       0.07851977      -0.16617998       0.00000000      -0.00045926
O       -1.49671077       0.00000001       0.07737963       0.16618026       0.00000000      -0.00039308
1
Lattice="10.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 10.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-49078.1951382926 free_energy=-49078.1951382926 pbc="T T T"
Zn       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000
1
Lattice="10.0 0.0 0.0 0.0 10.0 0.0 0.0 0.0 10.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-2034.82069902266 free_energy=-2034.81298590667 pbc="T T T"
O        0.00000000       0.00000000       0.00000000       0.00000000       0.00000000       0.00000000
```

**Comment by noambernstein** [comment]
I am actually surprised this was an issue. The Fortran reader should be
able to handle nonperiodic xyz files, and the corresponding neighbor lists
should be correct (I.e. no neighbors across images). I thought that’s what
it used to figure out if it had any isolated systems. It might be worth
another look at how the fitting code does this.

**Comment by gabor1** [comment]
If we could, we should fix it. But I fear that the cutoff is not known at read time. Maybe we need to make the Fortran atoms object fully support nonperiodic structures, not via a fake vacuum periodic cell

-- Gábor

**Comment by DGKang234** [comment]
If I may throw an idea what if GAP detected the nonperiodic structures then, place the nonperiodic structure in "a single simulation box (which has 2 or 3 times distance from the corner atom)". 
(if ._extxyz_ shows this it would be with **Lattice vector** component with **pbc= F F F**)

for example like this with pseudo vacuum box.
![Picture 1](https://user-images.githubusercontent.com/59916992/102694210-8819b800-4217-11eb-9074-b1fc4b6f15b2.png)

DG Kang

**Comment by jameskermode** [comment]
From memory I think the Fortran neighbour list does already support open boundaries in directions where pbc is False, as detected from zero lattice length in that direction, but one would need to check the code to be sure.

**Comment by gabor1** [comment]
Maybe it's just the "isolated atom detector" that isn't aware of non-pbc?

-- Gábor

**Comment by bernstei** [comment]
> On Dec 19, 2020, at 2:40 PM, gabor1 <notifications@github.com> wrote:
> 
> Maybe it's just the "isolated atom detector" that isn't aware of non-pbc?

Fortran is aware of nonperiodic configs, and the gap fit just uses the fortran neighbor list code, which should work.  I'm not sure how it handles a config _with_ a Lattice but with pbc="F F F".  I'm pretty confident that without a lattice (or with lattice vector = 0) it should be properly nonperiodic.  I'll take a look, since I see no good reason why it failed (and I might as well figure out if there's an issue, since there's some related workflow code that needs to be consistent).

**Comment by gabor1** [comment]
Ok. It's got to be related to how gap_fit decides to classify a config as "isolated atom"

-- Gábor

**Comment by bernstei** [comment]
On Dec 20, 2020, at 9:32 AM, gabor1 <notifications@github.com<mailto:notifications@github.com>> wrote:

Ok. It's got to be related to how gap_fit decides to classify a config as "isolated atom"

The answer is that it doesn't think those configs have an "energy" parameter, so it ignores them.  Looking into why right now.

Noam


____________
|
|
|
U.S. NAVAL
|
|
_RESEARCH_
|
LABORATORY

Noam Bernstein, Ph.D.
Center for Materials Physics and Technology
U.S. Naval Research Laboratory
T +1 202 404 8628  F +1 202 404 7546
https://www.nrl.navy.mil

**Comment by bernstei** [comment]
Aha - if you omit Lattice, it doesn't parse the "energy=<value>" parts. In fact I believe that it assumes that it's just not extended xyz, just plain xyz, although glancing at the C reader I'm not sure why it's not trying to parse other key-value pairs.  Yet another thing that will be dealt with once we standardize the extxyz format, I guess.

If you don't want to specify a large cell, you can specify Lattice="0 0 0   0 0 0   0 0 0", which will make the Fortran treat the config as nonperiodic, but you can't skip it completely.

**Comment by DGKang234** [comment]
@bernstei 
So for fitting with nonperiodic structure have to have Lattice component which is either 0 vectors or very big vectors.
Should nonperiodic .extxyz file still have to have "pbc= T T T", rather than pbc=F F F?

**Comment by gabor1** [comment]
I don't think that matters as long as the lattice vectors are huge. If you set lattice vectors as 0 then of course you need PBC to be FFF

-- Gábor

**Comment by noambernstein** [comment]
In this case the value of pbc basically doesn't matter.  Technically, if
you make the vectors _very_ big, and set pbc to T, it will try to make a
neighbor list, which may take a lot of memory (of order
(box_size/cutoff)^3), but unless you really go crazy with the box size it's
unlikely to be a practical issue.  But in either case the behavior should
be correct.
